cpp/src/gandiva/greedy_dual_size_cache.h

```diff
@@ -0,0 +1,150 @@
+    Key cache_key;
+  };
+
+ public:
+  struct hasher {
+    template <typename I>
+    std::size_t operator()(const I& i) const {
+      return i.Hash();
+    }
+  };
+  // a map from 'key' to a pair of Value and a pointer to the priority value
+  using map_type = std::unordered_map<
+      Key, std::pair<ValueCacheObject<Value>, typename std::set<PriorityItem>::iterator>,
+      hasher>;
+
+  explicit GreedyDualSizeCache(size_t capacity) : cache_capacity_(capacity) {
+    this->inflation_cost_ = 0;
+  }
+
+  GreedyDualSizeCache<Key, Value>() : cache_capacity_(capacity) {}
+
+  ~GreedyDualSizeCache() = default;
+
+  size_t size() const { return map_.size(); }
+
+  size_t capacity() const { return this->cache_capacity_; }
+
+  bool empty() const { return map_.empty(); }
+
+  bool contains(const Key& key) { return map_.find(key) != map_.end(); }
+
+  void insert(const Key& key, const ValueCacheObject<Value>& value) {
+    typename map_type::iterator i = map_.find(key);
+    // check if element is not in the cache to add it
+    if (i == map_.end()) {
+      // insert item into the cache, but first check if it is full, to evict an item
+      // if it is necessary
+      if (size() >= this->cache_capacity_) {
+        evict();
+      }
+
+      // insert the new item
+      auto iter = this->priority_set_.insert(
+          PriorityItem(value.cost + this->inflation_cost_, value.cost, key));
+      // save on map the value and the priority item iterator position
+      map_[key] = std::make_pair(value, iter.first);
+    }
+  }
+
+  arrow::util::optional<ValueCacheObject<Value>> get(const Key& key) {
+    // lookup value in the cache
+    typename map_type::iterator value_for_key = map_.find(key);
+    if (value_for_key == map_.end()) {
+      // value not in cache
+      return arrow::util::nullopt;
+    }
+    PriorityItem item = *value_for_key->second.second;
+    // if the value was found on the cache, update its cost (original + inflation)
+    if (item.actual_priority != item.original_priority + this->inflation_cost_) {
+      priority_set_.erase(value_for_key->second.second);
+      auto iter = priority_set_.insert(
+          PriorityItem(item.original_priority + this->inflation_cost_,
+                       item.original_priority, item.cache_key));
+      value_for_key->second.second = iter.first;
+    }
+    return value_for_key->second.first;
+  }
+
+  void clear() {
+    map_.clear();
+    priority_set_.clear();
+  }
+
+ private:
+  void evict() {
+    // evict item from the beginning of the set. This set is ordered from the
+    // lower priority value to the higher priority value.
+    typename std::set<PriorityItem>::iterator i = priority_set_.begin();
+    // update the inflation cost related to the evicted item
+    this->inflation_cost_ = (*i).actual_priority;
+    map_.erase((*i).cache_key);
+    priority_set_.erase(i);
+  }
+
+ private:
+  map_type map_;
+  std::set<PriorityItem> priority_set_;
+  int64_t inflation_cost_;
```

Comment:
uint64_t? (same as priority cost)
Also since the inflation_cost is always increasing, is there a possibility of overflow?